### PR TITLE
[Sema] Fix several redeclaration checking bugs

### PR DESCRIPTION
**Pull request by hamishknight**
Currently we only give subscripts and var decls custom overload types if they're in generic extensions. However, because we give them no custom overload type in any other case, we don't detect for example a conflict with a previous declaration in the body of the extended type – leading to accepting invalid and crashing in SILGen.

This commit changes the overload type logic such that properties and subscripts are always given custom overload types, which is determined by:

- The interface type of the decl (for subscripts only; as variables cannot be overloaded by type)
- The 'self' type of the context, if any
- The generic signature of the context, if any

Additionally, this commit adds a new `swift::conflicting` overload to ensure that different declarations always conflict even if their overload types are different.

Resolves [SR-7249](https://bugs.swift.org/browse/SR-7249), [SR-7250](https://bugs.swift.org/browse/SR-7250) & [SR-7251](https://bugs.swift.org/browse/SR-7251).

### CODE REVIEW

File: include/swift/AST/Decl.h
Comment (slavapestov) [inline include/swift/AST/Decl.h]: These could be bitfields for efficiency (I know this isn't part of your change but it might be easy to change it)

File: lib/AST/Decl.cpp
Comment (slavapestov) [inline lib/AST/Decl.cpp]: Nitpick: indent

**Review by slavapestov — COMMENTED** [review]
I think @DougGregor or @jrose-apple should look at this in more detail

File: lib/AST/Decl.cpp
Comment (hamishknight) [reply lib/AST/Decl.cpp]: Oops, thanks

File: include/swift/AST/Decl.h
Comment (hamishknight) [reply include/swift/AST/Decl.h]: Good idea, thanks

File: lib/AST/Decl.cpp
Comment (jrose-apple) [inline lib/AST/Decl.cpp]: What if the subscript itself is generic? I know that didn't work before either, but there should at least be a FIXME if you're going to redo the code without fixing it.

File: test/decl/overload.swift
Comment (jrose-apple) [inline test/decl/overload.swift]: Please don't remove this example; it may be relevant that two things are in the same pattern-binding. You can make a new struct instead if you want to test the other case.

File: test/decl/overload.swift
Comment (jrose-apple) [inline test/decl/overload.swift]: Nitpick: please capitalize starts of sentences in comments.

File: lib/AST/Decl.cpp
Comment (jrose-apple) [inline lib/AST/Decl.cpp]: Variables do not conflict with zero-argument functions (and you've implemented them to not conflict, so that's okay).

File: lib/Sema/TypeCheckDecl.cpp
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: This is going to redo the check from earlier. Is it possible to not do that?

File: lib/AST/Decl.cpp
Comment (jrose-apple) [reply lib/AST/Decl.cpp]: Why are these new checks here and not inside the original `conflicting`?

**Review by jrose-apple — COMMENTED** [review]
Thanks for fixing these! I'm a little suspect of some of the changes but I think the overall direction makes sense (in particular the important part where properties and subscripts now get sensible types).

File: lib/AST/Decl.cpp
Comment (hamishknight) [reply lib/AST/Decl.cpp]: @jrose-apple Variables do indeed conflict with zero argument functions (we have `mixed_redecl6` in the `overload.swift` test for that), e.g:

```
struct Foo {
  var i: Int { return 0 }
  func i() {} // error: Invalid redeclaration of 'i()'
}
```
---
These new checks are in this function as they're only relevant when comparing the overload types (they prevent us from saying that two decls don't conflict because the overload types differ).

To be honest, I'm not too mad on having this logic separated across three functions (the two `conflicting` overloads, and the logic that computes the overload types), but I couldn't really see a better way to do it given the two current use cases of the logic. Any ideas on how to better structure things are more than welcome :)

File: lib/Sema/TypeCheckDecl.cpp
Comment (hamishknight) [reply lib/Sema/TypeCheckDecl.cpp]: @jrose-apple I agree it's not ideal. We could pass a `bool` flag, something like `areSignaturesConflicting` – at the time I didn't really feel it was worth it given the logic being repeated was fairly lightweight, but am happy to do so if you feel it would be worthwhile.

File: lib/AST/Decl.cpp
Comment (hamishknight) [reply lib/AST/Decl.cpp]: @jrose-apple Looking into it, I think we already correctly handle generic subscripts in this PR. The generic signature just gets nested in the type when we curry it with 'self'.

For example, for this subscript:

```swift
struct Foo<U> {
  subscript<T>(_ t: T) -> U {
    fatalError()
  }
}
```

Its overload type is:
```
(generic_function_type escaping
  (input=paren_type
    (bound_generic_struct_type decl=main.(file).Foo
      (generic_type_param_type depth=0 index=0)))
  (output=generic_function_type escaping
    (input=paren_type
      (generic_type_param_type depth=1 index=0))
    (output=generic_type_param_type depth=0 index=0)
    ( generic_sig=<τ_0_0, τ_1_0>))
  ( generic_sig=<τ_0_0>))
```

I think this is okay for the purposes for an overload type. I'll go ahead and add some tests for generic subscripts though.

Comment (jrose-apple) [reply lib/AST/Decl.cpp]: I was pretty sure it still has to be a GenericFunctionType, though, which is something that *wouldn't* happen if the subscript is generic but the context is not.

(I'll reply to the other comments later.)

Comment (hamishknight) [reply lib/AST/Decl.cpp]: @jrose-apple Ah okay, that makes sense. So presumably we need to lift the subscript's generic signature up to the curried function's generic signature.

Actually [just found an existing function](https://github.com/apple/swift/blob/d8b355dbaebd157ba75e19fbc6ace87ef88e3e98/lib/Sema/CSRanking.cpp#L257) that curries a type with a given context's 'self', so can just reuse that for this (it doesn't use a MetatypeType for a static member, but we didn't really need that to begin with as static members don't have conflicting signatures with instance members).

So for this subscript:
```swift
struct Foo {
  subscript<T>(_ t: T) -> T {
    fatalError()
  }
}
```
its overload type is now:
```
(generic_function_type escaping
  (input=paren_type
    (struct_type decl=main.(file).Foo))
  (output=function_type escaping
    (input=paren_type
      (generic_type_param_type depth=0 index=0))
    (output=generic_type_param_type depth=0 index=0))
  ( generic_sig=<τ_0_0>))
```

which matches up with the overload type for a similar method decl, e.g:
```swift
struct Foo {
  func foo<T>(_ t: T) -> T {
    fatalError()
  }
}
```

I've gone ahead and pushed the revisions for your feedback – thanks for your patience!

File: lib/AST/Decl.cpp
Comment (jrose-apple) [reply lib/AST/Decl.cpp]: > Variables do indeed conflict with zero argument functions – we have `mixed_redecl6` in the `overload.swift` test for that

Aargh, right. You're not allowed to declare them, even though you can end up with that effect by importing other modules.

File: lib/AST/Decl.cpp
Comment (jrose-apple) [inline lib/AST/Decl.cpp]: We have this as a convenience predicate on ASTContext, `isSwiftVersionAtLeast`.

File: lib/AST/LookupVisibleDecls.cpp
Comment (jrose-apple) [inline lib/AST/LookupVisibleDecls.cpp]: No need to jump up to the DeclContext here; Decls also have a `getASTContext()`. (But it'd be fastest to get it from the module `M`, which always has a direct reference.)

File: lib/AST/Decl.cpp
Comment (jrose-apple) [inline lib/AST/Decl.cpp]: I still don't know about this. It might be simpler to just have a second function that says "given that the signatures overlap, do the types conflict?" and makes this into an assertion instead of a check. I'm going to defer to @DougGregor on this one (which will slow down the review, sorry).

**Comment by jrose-apple** [comment]
The other thing we could consider doing is *warning* in pre-Swift-5 modes that what they're doing is probably not a good idea.

**Comment by hamishknight** [comment]
@jrose-apple Good idea, thanks, I've gone ahead and made those changes (keeping the second `swift::conflicting` function chaining to the first for now though, will see what Doug Gregor thinks).

For the warning, I used an out-parameter instead of a return with 3 values in order to avoid burdening the caller in the case where they just want to know if there's a conflict between two decls in the current Swift version (which IMO should be the common case). Let me know what you think about that.

Thanks again for the feedback!

File: lib/AST/Decl.cpp
Comment (DougGregor) [inline lib/AST/Decl.cpp]: I like this staging via a tri-state a lot, thank you!

File: lib/AST/Decl.cpp
Comment (DougGregor) [inline lib/AST/Decl.cpp]: Very nice cleanup here, thank you.

File: include/swift/AST/Decl.h
Comment (DougGregor) [inline include/swift/AST/Decl.h]: Can we make this previous form of `conflicting` static? It looks like only the new `conflicting` calls it.

**Review by DougGregor — APPROVED** [review]
This looks fantastic! I'd love to get this merged now, and my one (minor) comment could be addressed in a follow-up commit.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: include/swift/AST/Decl.h
Comment (hamishknight) [reply include/swift/AST/Decl.h]: @DougGregor We do still call the original `conflicting` function [here](https://github.com/apple/swift/pull/15412/files#diff-32cb5387532d533d00872d725748a4d9L921) in order to initially discount decls with different signatures (before we then go to calculate their overload type) – which means that we'll end up calling the first `conflicting` twice if the signatures conflict.

Jordan raised his concerns over this redundant additional calling [here](https://github.com/apple/swift/pull/15412#discussion_r176475034) and [here](https://github.com/apple/swift/pull/15412#discussion_r176786739) – I wasn't too concerned given the logic is fairly lightweight; but it would be nice to refactor in such a way where that's not the case. Thoughts?

(I notice that `overload_swift4.swift` is failing on Linux; I'll look into that...)

**Comment by hamishknight** [comment]
Apologies, there was an issue where we could emit more than one redeclaration deprecation warning per decl which I somehow didn't catch when testing locally. We don't run into the same problem for re-declaration errors as we mark those decls as invalid, which prevents us from checking them again.

I've gone ahead and pushed a fix for that.

File: include/swift/AST/Decl.h
Comment (DougGregor) [reply include/swift/AST/Decl.h]: Ah, I'd missed the remaining call to `conflicting`. I'm not too concerned about the extra calls, because this is a fairly quick check.

**Comment by DougGregor** [comment]
@hamishknight , I wonder if your change fixes this example we just tripper over elsewhere:

```
struct A<Element> {
	var a: String {
		return "a1"
	}
}

extension A {
	var a: String {
		return "a2"
	}
}
```

**Review by DougGregor — APPROVED** [review]

**Comment by hamishknight** [comment]
@DougGregor Yes, that's [SR-7249](https://bugs.swift.org/browse/SR-7249) and is fixed by this PR (in that we now emit a redeclaration error for it). We have a test case for it [here](https://github.com/apple/swift/pull/15412/files#diff-77d426ccfae20f18ff28083b5bfc6d0eR442).

**Comment by DougGregor** [comment]
Ah hah, great!

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
drat, something conflicted before merge.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by hamishknight** [comment]
Thanks for resolving the merge conflict @DougGregor! Looks like this is good to go now.

**Comment by DougGregor** [comment]
Thank you! Want to do the honors of resolving the JIRAs, @hamishknight ?

**Comment by aschwaighofer** [comment]
This appears to have broken ReactiveCocoa in the source compatibility suite

**Comment by aschwaighofer** [comment]
https://ci.swift.org/job/swift-master-source-compat-suite/1527/console

**Comment by jrose-apple** [comment]
Failure logs: https://ci.swift.org/job/swift-master-source-compat-suite/1527/artifact/swift-source-compat-suite/

```
/Users/buildnode/jenkins/workspace-private/swift-master-source-compat-suite/project_cache/ReactiveCocoa/ReactiveCocoa/NSObject+Intercepting.swift:389:19: error: ambiguous use of 'numberOfArguments'
        let count = UInt(methodSignature.numberOfArguments!)
                         ^
/Users/buildnode/jenkins/workspace-private/swift-master-source-compat-suite/project_cache/ReactiveCocoa/ReactiveCocoa/ObjC+Messages.swift:38:6: note: found this candidate
        var numberOfArguments: UInt { get }
            ^
Foundation.NSMethodSignature:7:14: note: found this candidate
    open var numberOfArguments: Int { get }
             ^
```

Maybe we were filtering this out before? Or maybe they meant to use `as UInt`?

**Comment by hamishknight** [comment]
Here's a minimal example of the issue:

```swift
import Foundation

@objc protocol P {
  var i: UInt { get }
}

class C : NSObject {
  @objc var i: Int {
    return 0
  }
}

let x = (C() as AnyObject).i // compiles before patch, "ambiguous use of 'i'" error after
```
Presumably they meant to make the protocol requirement match the interface for `NSMethodSignature` (i.e make it `var numberOfArguments: Int { get }`) – that would fix the issue.

I'm not too familiar with the workings of the "ambiguous use of" diagnostic, but before the patch we would have given both `i` in the protocol and the class decl the same overload type (`CanType()`), whereas now they have different overload types.

So presumably we now see them as two different overloads and diagnose that there's not enough context to tell them apart (the correct behaviour IMO), rather than seeing them as the same overload and favouring the `Int` overload over the `UInt` one.

What course of action should we take here? Do we need to find a way to downgrade the ambiguity error to a warning until Swift 5 mode?

**Comment by aschwaighofer** [comment]
@clackary Has XFAILed ReactiveCocoa for now. https://bugs.swift.org/browse/SR-7299

**Comment by aschwaighofer** [comment]
@jrose-apple @DougGregor @hamishknight : In the short term what is the way forward? To do you intend to update ReactiveCocoa (or talk to the maintainers to do so) or should we revert until we have a fix: new behavior guarded by swift mode? I think we are trying hard not to have source compat projects disabled for too long ...

**Comment by jrose-apple** [comment]
Doug and I are leaning towards asking the ReactiveCocoa folks for an update. The old code would have been picking a property basically at random (whatever it saw first), and we can't think of a sensible way to downgrade this one to a warning.

**Comment by hamishknight** [comment]
@jrose-apple Do you know why `NSMethodSignature`'s `numberOfArguments` is being imported into Swift as an `Int` [despite being documented as being](https://developer.apple.com/documentation/foundation/nsmethodsignature/1519662-numberofarguments?language=objc) an `NSUInteger`, and therefore should (AFAIK) be imported as a `UInt`?

Just trying to figure out why the issue is only reproducible for them on iOS and tvOS, but not macOS.

**Comment by hamishknight** [comment]
@andersio has opened a PR here to fix things for ReactiveCocoa: https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3580

**Pull request merged**
